src/snowflake/connector/nanoarrow_cpp/ArrowIterator/DecimalConverter.cpp

```diff
@@ -0,0 +1,102 @@
+//
+// Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+//
+
+#include "Python/Common.hpp"
+#include "DecimalConverter.hpp"
+#include "Python/Helpers.hpp"
+#include <memory>
+#include <string>
+
+namespace sf
+{
+
+DecimalBaseConverter::DecimalBaseConverter()
+: m_pyDecimalConstructor(initPyDecimalConstructor())
+{
+}
+
+py::UniqueRef& DecimalBaseConverter::initPyDecimalConstructor()
+{
+  static py::UniqueRef pyDecimalConstructor;
+  if (pyDecimalConstructor.empty())
+  {
+    py::UniqueRef decimalModule;
+    py::importPythonModule("decimal", decimalModule);
+    py::importFromModule(decimalModule, "Decimal", pyDecimalConstructor);
+    Py_XINCREF(pyDecimalConstructor.get());
+  }
+
+  return pyDecimalConstructor;
+}
+
+DecimalFromIntConverter::DecimalFromIntConverter(ArrowArrayView* array,
+                                   int precision, int scale)
+  : m_array(array),
+    m_precision(precision),
+    m_scale(scale)
+  {
+  }
+
+PyObject* DecimalFromIntConverter::toPyObject(int64_t rowIndex) const
+{
+  if(ArrowArrayViewIsNull(m_array, rowIndex)) {
+    Py_RETURN_NONE;
+  }
+  int64_t val = ArrowArrayViewGetIntUnsafe(m_array, rowIndex);
+  py::UniqueRef decimal(
+        PyObject_CallFunction(m_pyDecimalConstructor.get(), "L", val));
+  return PyObject_CallMethod(decimal.get(), "scaleb", "i", -m_scale);
+}
+
+NumpyDecimalConverter::NumpyDecimalConverter(ArrowArrayView* array,
+                                 int precision, int scale, PyObject * context)
+  : m_array(array),
+    m_precision(precision),
+    m_scale(scale),
+    m_context(context)
+  {
+  }
+
+PyObject* NumpyDecimalConverter::toPyObject(int64_t rowIndex) const
```

Comment:
This also seems new. When will we use the numpy version of converter?

Reply:
it's in the old arrow converter, but just defined in the `DecimalConverter.hpp` file, I reorganized the code to make it more clear.

there is a config called `self._use_numpy` defined in the python connection, it will be passed all way down to the c arrow converter